src/firebase-app.ts

```diff
@@ -57,129 +59,80 @@ export interface FirebaseAccessToken {
-          const hasExpirationChanged = (this.cachedToken_ && this.cachedToken_.expirationTime !== token.expirationTime);
-          if (!this.cachedToken_ || hasAccessTokenChanged || hasExpirationChanged) {
-            this.cachedToken_ = token;
-            this.tokenListeners_.forEach((listener) => {
-              listener(token.accessToken);
-            });
-          }
-
-          // Establish a timeout to proactively refresh the token every minute starting at five
-          // minutes before it expires. Once a token refresh succeeds, no further retries are
-          // needed; if it fails, retry every minute until the token expires (resulting in a total
-          // of four retries: at 4, 3, 2, and 1 minutes).
-          let refreshTimeInSeconds = (result.expires_in - (5 * 60));
-          let numRetries = 4;
-
-          // In the rare cases the token is short-lived (that is, it expires in less than five
-          // minutes from when it was fetched), establish the timeout to refresh it after the
-          // current minute ends and update the number of retries that should be attempted before
-          // the token expires.
-          if (refreshTimeInSeconds <= 0) {
-            refreshTimeInSeconds = result.expires_in % 60;
-            numRetries = Math.floor(result.expires_in / 60) - 1;
-          }
-
-          // The token refresh timeout keeps the Node.js process alive, so only create it if this
-          // instance has not already been deleted.
-          if (numRetries && !this.isDeleted_) {
-            this.setTokenRefreshTimeout(refreshTimeInSeconds * 1000, numRetries);
-          }
-
-          return token;
-        })
-        .catch((error) => {
-          let errorMessage = (typeof error === 'string') ? error : error.message;
-
-          errorMessage = 'Credential implementation provided to initializeApp() via the ' +
-            '"credential" property failed to fetch a valid Google OAuth2 access token with the ' +
-            `following error: "${errorMessage}".`;
-
-          if (errorMessage.indexOf('invalid_grant') !== -1) {
-            errorMessage += ' There are two likely causes: (1) your server time is not properly ' +
-            'synced or (2) your certificate key file has been revoked. To solve (1), re-sync the ' +
-            'time on your server. To solve (2), make sure the key ID for your key file is still ' +
-            'present at https://console.firebase.google.com/iam-admin/serviceaccounts/project. If ' +
-            'not, generate a new key file at ' +
-            'https://console.firebase.google.com/project/_/settings/serviceaccounts/adminsdk.';
-          }
-
-          throw new FirebaseAppError(AppErrorCodes.INVALID_CREDENTIAL, errorMessage);
-        });
-
-      return this.cachedTokenPromise_;
+  public getToken(forceRefresh = false): Promise<FirebaseAccessToken> {
```

Comment:
Can the RTDB component just set `forceRefresh` to true if the expiration time has been hit? That way, we only need to deal with TOKEN_REFRESH_THRESHOLD_MILLIS in one place.

Reply:
These 2 constants are used for slightly different purposes:

1. In firebase-app it is used to determine how early we should treat a valid token as expired.
2. In database-internal it is used to determine when to schedule the next refresh event.

These don't have to be the same value. For instance, we can have FirebaseApp invalidate a token 5 minutes before it expires, but have RTDB trigger the token refresh 10 minutes before the current token expires.

To make the above distinction clear, I'm renaming the constant in firebase-app to `TOKEN_EXPIRY_THRESHOLD_MILLIS`. This constant doesn't have anything to do with refreshing tokens. It's simply when we treat the token as expired.